### ISSUE TITLE
RPM should express scriptlet dependencies

### ISSUE THREAD

**Issue by AdmiralNemo** [elastic/elasticsearch#24397]
**Elasticsearch version**: 5.3.1

**OS version**: CentOS 7

**Description of the problem including expected versus actual behavior**:

The _elasticsearch_ RPM needs to express the pre/post-(un)install scriptlet dependencies. Without them, it is possible that the package manager can attempt to install the package before the commands used by the scripts are available, causing the scripts to fail:

```
sudo grep 'command not found' /var/log/anaconda/packaging.log
16:06:14,152 INFO packaging: /var/tmp/rpm-tmp.PPr6cW: line 64: groupadd: command not found
16:06:14,152 INFO packaging: /var/tmp/rpm-tmp.PPr6cW: line 71: useradd: command not found
16:06:14,153 INFO packaging: /var/tmp/rpm-tmp.82AXnf: line 119: chown: command not found
16:06:14,153 INFO packaging: /var/tmp/rpm-tmp.82AXnf: line 120: chown: command not found
16:06:14,153 INFO packaging: /var/tmp/rpm-tmp.82AXnf: line 121: chown: command not found
16:06:14,153 INFO packaging: /var/tmp/rpm-tmp.82AXnf: line 122: chmod: command not found
16:06:14,153 INFO packaging: /var/tmp/rpm-tmp.82AXnf: line 123: chmod: command not found
16:06:14,153 INFO packaging: /var/tmp/rpm-tmp.82AXnf: line 126: chmod: command not found
```

This typically happens when installing _elasticsearch_ via kickstart, since everything is done in a single transaction.

Here are the dependencies needed (for RHEL/CentOS, other RPM-based distributions are likely to have different package names):

```
Requires(pre): shadow-utils
Requires(post): coreutils, systemd
Requires(preun): systemd
Requires(postun): coreutils, shadow-utils
```

**Comment by spinscale** [comment]
as other distributions have different package names and we are only building a single RPM, we cannot have those as hard dependencies, unless we build a package per distribution, or am I missing something here?

**Comment by jasontedor** [comment]
I agree with @spinscale (plus systemd is not CentOS 6 but is CentOS 7 adding more complexity), this is too confusing for users. Also, these packages are almost always installed already where needed. I think the packaging is fine as is.

**Comment by AdmiralNemo** [comment]
You can use paths to binaries instead of package names to achieve better cross-distribution support.

As I indicated in the description, these tools are not "almost always" installed when everything is done in a single transaction. In fact, because the _elasticsearch_ package expresses no dependencies at all, it gets installed extremely early, so it certainly isn't "fine as is."

**Comment by fpoirotte** [comment]
I agree with @AdmiralNemo: because most of the dependencies are missing from the packaging, installing ElasticSearch from a kickstart installation is next to impossible.

For example, the package for ES 6 is also missing a dependency on the JVM, and thus also fails to install if java is not already installed (be it the Sun JVM or an alternative JVM like OpenJDK) because the preinstall scriptlet now has this check:
```
# Check for these at preinst time due to failures in postinst if they do not exist
if [ -x "$JAVA_HOME/bin/java" ]; then
    JAVA="$JAVA_HOME/bin/java"
elif command -v java; then
    JAVA=`command -v java`
else
    JAVA=""
fi

if [ -z "$JAVA" ]; then
    err_exit "could not find java; set JAVA_HOME or ensure java is in PATH"
fi
```
I would expect the ES package to have a dependency on the "java" virtual package provided by both the OpenJDK implementation and the reference (Sun/Oracle) implementation.

The fact that different distributions use different package names is a pity, but you can easily workaround that problem by adding conditionals, eg.
```
%if 0%{?rhel}
Requires: java >= 1:1.6.0 # or whatever the minimum version should be
%endif
```

This is what the Fedora project has been using for years to keep a single .spec file that works on both Fedora-based distributions and RHEL-based distributions. See https://fedoraproject.org/wiki/Packaging:DistTag#Conditionals for more information on that.

**Comment by jasontedor** [comment]
@fpoirotte Since Elasticsearch 7, we bundle a JDK with the Elasticsearch distribution. Among many reasons, one is to address this dependency problem in a way that is simple for our users.

**Comment by fpoirotte** [comment]
Thank you for your prompt answer.

Also, good to hear the situation has been taken care of for ES 7 (altough I personnaly dislike the idea of a 330 MB download...)

Are there any plans to fix this for ElasticSearch 6 too?
As an aside, how do you plan to keep up with security updates related to the JDK? Do you plan to release a new patch release of elasticsearch each time a new CVE for Java gets fixed?

**Comment by jasontedor** [comment]
@fpoirotte Good questions.

> Are there any plans to fix this for ElasticSearch 6 too?

We do not plan to make any changes in the Elasticsearch 6.x series related to this.

> As an aside, how do you plan to keep up with security updates related to the JDK? Do you plan to release a new patch release of elasticsearch each time a new CVE for Java gets fixed?

We will handle these on a case-by-case basis, as we do with any other dependency. The vast majority of the CVE against Java are in components that Elasticsearch is not even exposed to, so they wouldn't present security risks against Elasticsearch. To date, no CVE in Java has impacted Elasticsearch. We also have layers of defense with the security manager and seccomp, limiting the blast radius of any CVE there. Should there be a case that there is a CVE present in Java that impacts Elasticsearch, we would treat it like we treat any other vulnerability against Elasticsearch.
